### PR TITLE
Make all FFI _impl helpers safe rust

### PR DESCRIPTION
**Pull request by scovich**
Several FFI entry points are split into a pair of function, where the "inner" function does the actual work and returns a `DeltaResult` which the "outer" function simply converts to `ExternResult`:
```rust
pub unsafe extern "C" fn foo(
      ... args ...,
    allocate_error: AllocateErrorFn,
) -> ExternResult<*mut Foo> {
    foo_impl(... args...).into_extern_result(allocate_error)
}
 
unsafe fn foo_impl(
      ... args ...,
    allocate_fn: AllocateErrorFn,
 ) -> DeltaResult<*mut Foo> {
   let foo = unsafe { ... args ... };
    ... do the actual work ...
}
```
This PR observes that the "outer" function can _also_ capture all the unsafe code, so that the "inner" function becomes normal safe rust.

### CODE REVIEW

File: ffi/src/scan.rs [hunk 1]
```diff
@@ -295,7 +296,8 @@ pub unsafe extern "C" fn get_from_map(
     key: KernelStringSlice,
     allocate_fn: AllocateStringFn,
 ) -> NullableCvoid {
-    let string_key = String::try_from_slice(key);
+    // TODO: Return ExternResult to caller instead of panicking?
+    let string_key = String::try_from_slice(key).unwrap();
```
Comment (nicklan) [inline ffi/src/scan.rs:300]:
```suggestion
    let string_key = unsafe { String::try_from_slice(key) }.unwrap();
```

**Review by nicklan — APPROVED** [review]
nice! just one tiny suggestion, but lgtm

File: ffi/src/scan.rs [hunk 2]
```diff
@@ -295,7 +296,8 @@ pub unsafe extern "C" fn get_from_map(
...
     map.values
         .get(&string_key)
         .and_then(|v| allocate_fn(v.as_str().into()))
```
Comment (scovich) [inline ffi/src/scan.rs:303]: aside: I'm pretty sure we don't need the `.as_str()` there?

File: ffi/src/scan.rs (hunk 1 above)
Comment (scovich) [reply ffi/src/scan.rs:300]: heh. missed one.

**Review by roeap — APPROVED** [review]
👍

**Pull request merged**